src/shogun/kernel/PeriodicKernel.h

```diff
@@ -0,0 +1,180 @@
+
+		/** initialize kernel
+		 *
+		 * @param l features of left-hand side
+		 * @param r features of right-hand side
+		 * @return if initializing was successful
+		 */
+		virtual bool init(CFeatures* l, CFeatures* r);
+
+		/** clean up kernel */
+		virtual void cleanup();
+
+		/** return what type of kernel we are
+		 *
+		 * @return kernel type PERIODIC
+		 */
+		virtual EKernelType get_kernel_type() { return K_PERIODIC; }
+
+		/** return the kernel's name
+		 *
+		 * @return name PeriodicKernel
+		 */
+		virtual const char* get_name() const { return "PeriodicKernel"; }
+
+		/** set the kernel's width
+		 *
+		 * @param w kernel width
+		 */
+		virtual void set_width(float64_t w)	{ width=w; }
+
+		/** return the kernel's width
+		 *
+		 * @return kernel width
+		 */
+		virtual float64_t get_width() const	{ return width;	}
+
+		/** set the kernel's period
+		 *
+		 * @param w kernel period
+		 */
+		virtual void set_period(float64_t p)	{ period=p; }
+
+		/** return the kernel's period
+		 *
+		 * @return kernel perid
+		 */
+		virtual float64_t get_period() const	{ return period;	}
+
+		/** return derivative with respect to specified parameter
+		 *
+		 * @param param the parameter
+		 * @param index the index of the element if parameter is a vector
+		 *
+		 * @return gradient with respect to parameter
+		 */
+		virtual SGMatrix<float64_t> get_parameter_gradient(
+				const TParameter* param, index_t index=-1);
+
+	protected:
+		/** compute kernel function for features a and b
+		 * idx_{a,b} denote the index of the feature vectors
+		 * in the corresponding feature object
+		 *
+		 * @param idx_a index a
+		 * @param idx_b index b
+		 * @return computed kernel function at indices a,b
+		 */
+		virtual float64_t compute(int32_t idx_a, int32_t idx_b);
+
+
+		/** compute the euclidean distance between features a and b
+		 * idx_{a,b} denote the index of the feature vectors
+		 * in the corresponding feature object.
+		 *
+		 * @param idx_a index a
+		 * @param idx_b index b
+		 * @return computed the distance
+		 *
+		 * Note that this function is very similar the distance function in
+		 * CGaussianKernel. However, this function computes the standard
+		 * euclidean distance without any factors or squaring unlike the one
+		 * in CGaussianKernel.
+		 */
+		virtual float64_t distance(int32_t idx_a, int32_t idx_b);
+	private:
+		/** helper function to compute quadratic terms in
+		 * (a-b)^2 (== a^2+b^2-2ab)
+		 */
+		void precompute_squared();
+
+		/** helper function to compute quadratic terms in
+		 * (a-b)^2 (== a^2+b^2-2ab)
+		 *
+		 * @param buf buffer to store squared terms (will be allocated)
+		 * @param df dot feature object based on which k(i,i) is computed
+		 * */
+		void precompute_squared_helper(float64_t* &buf, CDotFeatures* df);
+
+		void init();
+
+	protected:
+		/** width */
+		float64_t width;
+		/** period */
+		float64_t period;
+		/** squared left-hand side */
+		float64_t* sq_lhs;
```

Comment:
This comes from old code. We are now using SGVector (no pointer, this is on the stack). It automatically handles memory. No need for any SG_FREE SG_MALLOC anymore